src/SessionManager.cc

```diff
@@ -165,15 +138,15 @@ void NetSessions::ProcessTransportLayer(double t, const Packet* pkt, size_t rema
 
```

Comment:
One of the major reasons this is a draft PR is because of a performance issue right here. `BuildConnIDKey()` generates a hash of the `ConnIDKey` object. It does this for every single packet received that passes through this code, which results in about a 1.5% performance loss just from the hashing. It's a big enough hit that it actually shows up in a flamegraph. I've considered a couple of fixes for this but neither of them felt like good choices:
- Cache the hashing result in some fashion in a map between `ConnIDKey` and the resulting hash. This means I could look it up and only do the hash if it's something that hasn't been seen before. This breaks the genericness of `SessionManager` and hides that there's a potential performance hit here from other implementers of new session types. It also needs some mechanism for cleaning up this cache periodically.
- Store `Connection`s separately from other sessions, still keyed by `ConnIDKey` instead of the hash. This covers the "common" case where most traffic is a session, but it again sort of breaks the whole point of `SessionManager` and `Session` being a generic way to store these

Reply:
So we're essentially hashing twice currently, right? (1) We hash `ConnID` into `hash_t`, and then (2) store that `hash_t` into the `SessionMap` (which technically doesn't *really* hash, because it's a map, but overhead should be similar to hashing; could also use `unordered_map` here).

Could we skip the first hash? Can we replace the `hash_t` with just a binary blob (`class KeyBlob`) instead and then put that into the `SessionMap`? For a connection the blob could be just be the raw `ConnID` data. This would make the `SessionMap` management more expensive (because the keys are larger) but it would remove prior hash on the ConnID, which is more expensive overall presumably.

There's a down-side: we'd no longer be using the cryptograph hash anymore that's currently in place (SipHash), potentially reopening an old vulnerability. But maybe `std::map` already takes care of that by using a tree (i..e., not hashing)? Not sure immediately.

Reply:
I don't think the overhead of the map is significant, in that it doesn't even show up in a Flamegraph. On the other hand, the hashing prior to storage shows up as ~2.5% of the runtime reading the 2009-M57-day11-18 file.

Reply:
> Can we replace the hash_t with just a binary blob (class KeyBlob) instead and then put that into the SessionMap? For a connection the blob could be just be the raw ConnID data. 

I think the problem with this idea is that the blob would have be a fixed size for `map`/`unordered_map` to use it, which then means I have to set some sort of maximum size. This gets into the question of whether anyone would ever design a session key that's bigger than the maximum.

Reply:
> I don't think the overhead of the map is significant. The insertion is about 0.03% of the runtime reading the 2009-M57-day11-18 file. On the other hand, the hashing prior to storage shows up as ~2.5%.

Yeah, right. That's why I was  trying to get rid of the first hash and move all the work into the map.

> I think the problem with this idea is that the blob would have be a fixed size for `map`/`unordered_map` to use it, which then means I have to set some sort of maximum size.

I don't think it would need to. If we wrap this into our own class with a custom comparison operator (or hash computation for `unordered_map`), we could support binary data of any size.

Reply:
I'm fairly sure you can't do that. If you create a `std::map<Base, Value>`, any comparison during insertion through that map will use a version of the comparator that takes two `Base` objects. That's the case even if you're inserting child classes to `Base`. Calls to `find()` looking for child objects does do comparisons against `Base` and the child type. You can `static_cast` the `Base` object to be a child object, but you have to implement comparators both for A vs B and B vs A. See https://godbolt.org/z/joqYfccbq for example.

Reply:
I think we are talking about different things, I didn't mean to use inheritance. Let's chat about it. :)

Reply:
> There's a down-side: we'd no longer be using the cryptograph hash anymore that's currently in place (SipHash), potentially reopening an old vulnerability.
From my understanding the cryptographically hashed UIDs are something different. UIDs are intended to uniquely identify a (finished) connection in the logs. In the sessionization context we just look at the currently open connections, which can be identified using the 5-tuple.

Reply:
> From my understanding the cryptographically hashed UIDs are something different.

it's two different things: here, the concern was that collisions in the session table could lead to bad performance if the wrong type of input traffic comes along. Zeek used to be vulnerable to that: there's an old paper that attacks our implemtnastion at the time to cause DOS with the volume of a modem line (nice paper!). We originally fixed that with a cryptographic hash that made the hash behavior non-predictable—and that's the hashing that the code on this PR ends up using by relying on `HashKey`.

That said, I had forgotten that this is not an issue anymore: by using `std::map` for the session table these days, we're not relying on hashing anymore, but on trees that don't have that vulnerability to begin with. So I think it's indeed safe to get rid of that first hash altogether.